### DISCUSSION TITLE
Injecting interface implementations with `@Inject` without a provider.

### DISCUSSION THREAD

**Discussion by dave08** [evant/kotlin-inject#135] [Ideas]
Currently, using `@Inject` on the implementation of an interface only provides the implementation to the component, but if the interface needs to be resolved we need to write a `@Provides` as a kind of alias. In frameworks like Micronaut, the DI knows how to automatically resolve the interface if there's only one implmentation. If there are multiple implementations, if one of them in `@Primary`, that one is resolved, and if another class requires all the implementations to be injected it can get them by `List<SomeInterface>` and all implementations are injected.

For testing, there's a `@Replaces(SomeClass)` annotation to override any primary implementation.

See: https://docs.micronaut.io/latest/guide/#replaces

**Comment by evant** [comment]
I really don't know how you'd go about finding if there's more than one implementation of an interface across a project in an efficient way.

**Comment by vRallev** [comment]
This is the solution to this problem: https://github.com/amzn/kotlin-inject-anvil?tab=readme-ov-file#contributesbinding
